### PR TITLE
menucmd.c: Prevent infinite recursion in menu_toogle_timer()

### LINKED ISSUE #43: Infinite recursion until segfault starting timer after -DGSEAL_ENABLE
Commit 3c19c4c0 enabled compilation with -DGSEAL_ENABLE, which required some changes. Following these changes, it's impossible to start the timer. Trying to do so results in infinite recursion and eventually segfault.

The cause is this change in menus.c:
https://github.com/GnoTime/gnotime/commit/3c19c4c0d2c3e0fa62f187acb957e7707595dfa4#diff-ad990f8c2d8bd10fd16c3a68e26e1c003fe3f6609b8abfd508422cf9096b7577R285-R290

In fact, the comment already in place there explains what happens.

CC @markuspg

### PR DESCRIPTION
**Pull request by oskarb**
menu_toogle_timer() now considers the current (new) state of the check menu item as the desired state, and will only trigger a timer state change if the current state is wrong. This makes it a no-op if the state is already correct, preventing infinite recursion when the handler is called in response to code setting the menu item state.

Fixes #43.

### CODE REVIEW

**Comment by oskarb** [comment]
This fixes the problem. Though I also question why we even have so many menu items.

**Review by markuspg — DISMISSED** [review]
Thank you for coming up with a fix for the regression introduced by me!

Did you test activating the timer just by means of the menu item or by the toolbar button too? For me the toolbar button is without effect, because it fails to be casted to a `GtkCheckMenuItem`. Maybe this could be fixed simply by attempting two casts (either `GtkCheckMenuItem` or `GtkToggleButton`)  and taking the state from the successful cast?

With the menu item I get further, but the code still fails on my machine - apparantely due to Wayland. Although I'm puzzled, because my machine is not Wayland-based since yesterday and I'm using the project timer frequently.

**Comment by oskarb** [comment]
I tested a bunch of different things, like double-click in the project tree, the different menu options, and the active/idle dialog, but it looks like I missed the toolbar button.

While it's possible for a toolbar button to be a toggle button (that would contain state and emit a toggled event), it seems we [currently use just a regular button](https://github.com/GnoTime/gnotime/blob/a14c15444f6596d7f4559dccf7ef0ab30603c8db/src/toolbar.c#LL201C5-L204C15). As long as we stick to that, this button does not suffer from the recursion issue, and the event handler for that could use the old logic.

I added a check for widget type in the handler. We could have used a separate handler, but there is some overlap and overall still not too complicated so I think this should be a good trade-off.

**Comment by oskarb** [comment]
Btw, there is an issue here in that if the menu toggle item is used to *start* the timer but no project is selected, then the timer will of course not start, but the menu item will still have the "active" state. I think this is an existing issue and I didn't want to complicate this code by trying to reset the state. A better future fix for that is probably to disable the menu item when no project is either running or selected.

**Comment by oskarb** [comment]
@markuspg How do you feel about the updated version of this? I cannot merge it without another approval. After this is done I can test and finalize #40 which is about removing the last GtkOptionMenu.

After that I am thinking to see if any of the glade files are trivially convertible to GtkBuilder, before taking a stab on the more challenging one (e.g. GnomePropertyBox in the preferences dialog).

**Comment by markuspg** [comment]
It still crashes, but unfortunately I don't get a meaningful stacktrace. I'm growing weary of these old and broken frameworks ;-) What would you think about bumping the dependencies to _GTK 3_ and _Debian 10 "Buster"_ and commenting out any not-working code for the time being? I wonder if this would be more productive rather than trying to make things work with these old crutches.

A thing I would still like to realize before is the `GnomeEntry` and `GnomePropertyDialog` situation.

`GnomeEntry` I successfully merged into our codebase (see the branch `markuspg/dev` as `GttEntry`. It is as "broken" as `GnomeEntry`. It should work, but the proper signal to activate the saving of the entries is not being emitted. That's why the current `GnomeEntry` widgets don't have history too. I have some slight tendency to go with "my" `GttEntry`, but I would be fine with replacing `GnomeEntry` by plain `GtkEntry` widgets too.

An import of `GnomePropertyDialog` as `GttPropertyDialog` I have still lying around somewhere. If I remember correctly this was smaller than `GttEntry` and works as intended. The added value of `GnomePropertyDialog` is litte, but I would like to keep it.

A further big item is the menu situation. I think for now I would comment out this code, transition to clean GTK 3 and then see what's the proper way to handle it there.

The idle timer will be a topic too. Right now it works with X only. So we either have to adjust it to work with Wayland too or to see what we do with this code. But as a first step I would port it to GTK 3 as is.

**Comment by markuspg** [comment]
I forgot the main point. Even if the fix is not 100% correct yet, I would still propose to merge it. To me it's more relevant not to lose existing features in the migration process than to avoid introducing bugs. When we are in GTK 3 land we'll anyhow conduct a thorough code review I guess, so I wouldn't want to slow us down by striving for maximum correctness (which we should do of course in the long run).

**Review by markuspg — APPROVED** [review]

**Pull request merged**
